iconfetcher.go

```diff
@@ -0,0 +1,108 @@
+package jujusvg
+
+import (
+	"io/ioutil"
+	"net/http"
+	"sync"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v5-unstable"
+)
+
+// An IconFetcher provides functionality for retrieving icons for the charms
+// within a given bundle.  The FetchIcons function accepts a bundle, and
+// returns a dict mapping charm paths to icon data.
+type IconFetcher interface {
+	FetchIcons(*charm.BundleData) (map[string]string, error)
+}
+
+// HttpFetcher is a default implementation of IconFetcher which retrieves charm
+// icons from the web using the URL generated by IconURL on that charm.  The
+// HTTP Client used may be overridden by an instance of http.Client.  The icons
+// may optionally be fetched concurrently.
+type HttpFetcher struct {
+	FetchConcurrently bool
+	IconURL           func(*charm.Reference) string
+	Client            *http.Client
+}
+
+// FetchIcons retrieves icon SVGs over HTTP.  If specified in the struct, icons
+// will be fetched concurrently.
+func (h *HttpFetcher) FetchIcons(b *charm.BundleData) (map[string]string, error) {
+	// If a client was not provided on creation, set a default.
+	if h.Client == nil {
+		h.Client = new(http.Client)
+	}
+
+	// Maintain a list of icons that have already been fetched.
+	alreadyFetched := make(map[string]bool)
+
+	// WaitGroup and channels for concurrent fetching.
+	var wg sync.WaitGroup
+	errors := make(chan error, len(b.Services))
+
+	// Build the map of icons.
+	icons := make(map[string]string)
+	for _, serviceData := range b.Services {
+		charmId, err := charm.ParseReference(serviceData.Charm)
+		if err != nil {
+			return nil, errgo.Notef(err, "cannot parse charm %q", serviceData.Charm)
+		}
+		path := charmId.Path()
+
+		// Don't duplicate icons in the map.
+		if !alreadyFetched[path] {
+			alreadyFetched[path] = true
+
+			// Fetch concurrently or not, depending on option.
+			if h.FetchConcurrently {
```

Comment:
I'm slightly worried this could be unbounded, in the case of sensible bundles it's probably not a problem. I'd be inclined to make sure I had a maximum number of running go routines at a time. If that could also be one then you wouldn't need to check h.FetchConcurrently so often. I'm not sure though.